### PR TITLE
solution

### PR DESCRIPTION
**Pull request by GlebPashko**

### CODE REVIEW

File: src/main/java/mate/academy/hibernate/relations/dao/impl/ActorDaoImpl.java [hunk 1]
```diff
@@ -12,11 +15,36 @@ public ActorDaoImpl(SessionFactory sessionFactory) {
...
+            if (transaction != null) {
+                transaction.rollback();
+            }
+            throw new DataProcessingException("Error adding actor", e);
+        } finally {
+            if (session != null) {
+                session.close();
+            }
+        }
+
```
Comment (MentorMateAcademy) [inline src/main/java/mate/academy/hibernate/relations/dao/impl/ActorDaoImpl.java:36]:
```suggestion
```

File: src/main/java/mate/academy/hibernate/relations/dao/impl/CountryDaoImpl.java [hunk 2]
```diff
@@ -12,11 +15,36 @@ public CountryDaoImpl(SessionFactory sessionFactory) {
...
+            if (transaction != null) {
+                transaction.rollback();
+            }
+            throw new DataProcessingException("Error adding country", e);
+        } finally {
+            if (session != null) {
+                session.close();
+            }
+        }
+
```
Comment (MentorMateAcademy) [inline src/main/java/mate/academy/hibernate/relations/dao/impl/CountryDaoImpl.java:36]:
```suggestion
```

File: src/main/java/mate/academy/hibernate/relations/model/Actor.java [hunk 3]
```diff
@@ -1,8 +1,21 @@
...
+import jakarta.persistence.Table;
+
+@Entity
+@Table(name = "actors")
 public class Actor implements Cloneable {
+    @Id
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
     private Long id;
     private String name;
+
```
Comment (MentorMateAcademy) [inline src/main/java/mate/academy/hibernate/relations/model/Actor.java:17]:
```suggestion
```

File: src/main/java/mate/academy/hibernate/relations/model/Movie.java [hunk 4]
```diff
@@ -1,11 +1,28 @@
...
 import java.util.List;
 
+@Entity
+@Table(name = "movies")
 public class Movie implements Cloneable {
+    @Id
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
     private Long id;
     private String title;
+
```
Comment (MentorMateAcademy) [inline src/main/java/mate/academy/hibernate/relations/model/Movie.java:21]:
```suggestion
```

File: src/main/java/mate/academy/hibernate/relations/service/impl/ActorServiceImpl.java [hunk 5]
```diff
@@ -1,16 +1,26 @@
...
+import mate.academy.hibernate.relations.dao.impl.ActorDaoImpl;
 import mate.academy.hibernate.relations.model.Actor;
 import mate.academy.hibernate.relations.service.ActorService;
+import org.hibernate.SessionFactory;
 
 public class ActorServiceImpl implements ActorService {
+    private ActorDao actorDao;
+
+    public ActorServiceImpl(SessionFactory sessionFactory) {
+        actorDao = new ActorDaoImpl(sessionFactory);
```
Comment (MentorMateAcademy) [inline src/main/java/mate/academy/hibernate/relations/service/impl/ActorServiceImpl.java:13]: You don`t need sessionFactory in service layer implement this in dao

**Review by MentorMateAcademy — CHANGES_REQUESTED** [review]

File: src/main/java/mate/academy/hibernate/relations/dao/impl/ActorDaoImpl.java [hunk 6]
```diff
@@ -12,11 +15,35 @@ public ActorDaoImpl(SessionFactory sessionFactory) {
...
     @Override
     public Actor add(Actor actor) {
-        return null;
+        Session session = null;
+        Transaction transaction = null;
+
+        try {
+            session = factory.openSession();
+            transaction = session.beginTransaction();
+            session.save(actor);
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/hibernate/relations/dao/impl/ActorDaoImpl.java:24]:
```suggestion
            session.persist(actor);
```

File: src/main/java/mate/academy/hibernate/relations/dao/impl/ActorDaoImpl.java [hunk 7]
```diff
@@ -12,11 +15,35 @@ public ActorDaoImpl(SessionFactory sessionFactory) {
...
+        }
+        return actor;
     }
 
     @Override
     public Optional<Actor> get(Long id) {
-        return null;
+        Actor actor = null;
+        try (Session session = factory.openSession()) {
+            actor = session.get(Actor.class, id);
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/hibernate/relations/dao/impl/ActorDaoImpl.java:43]:
```suggestion
            return Optional.ofNullable(session.get(Actor.class, id));
```

File: src/main/java/mate/academy/hibernate/relations/dao/impl/ActorDaoImpl.java [hunk 8]
```diff
@@ -12,11 +15,35 @@ public ActorDaoImpl(SessionFactory sessionFactory) {
...
+                session.close();
+            }
+        }
+        return actor;
     }
 
     @Override
     public Optional<Actor> get(Long id) {
-        return null;
+        Actor actor = null;
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/hibernate/relations/dao/impl/ActorDaoImpl.java:41]:
```suggestion
```

File: src/main/java/mate/academy/hibernate/relations/dao/impl/ActorDaoImpl.java [hunk 9]
```diff
@@ -12,11 +15,35 @@ public ActorDaoImpl(SessionFactory sessionFactory) {
...
+        try (Session session = factory.openSession()) {
+            actor = session.get(Actor.class, id);
+        } catch (Exception e) {
+            throw new DataProcessingException("Error getting actor", e);
+        }
+        return Optional.ofNullable(actor);
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/hibernate/relations/dao/impl/ActorDaoImpl.java:47]:
```suggestion
```

File: src/main/java/mate/academy/hibernate/relations/dao/impl/ActorDaoImpl.java [hunk 10]
```diff
@@ -12,11 +15,35 @@ public ActorDaoImpl(SessionFactory sessionFactory) {
...
     }
 
     @Override
     public Optional<Actor> get(Long id) {
-        return null;
+        Actor actor = null;
+        try (Session session = factory.openSession()) {
+            actor = session.get(Actor.class, id);
+        } catch (Exception e) {
+            throw new DataProcessingException("Error getting actor", e);
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/hibernate/relations/dao/impl/ActorDaoImpl.java:45]:
```suggestion
            throw new DataProcessingException("Can't get Actor by id: " + id);
```

File: src/main/java/mate/academy/hibernate/relations/dao/impl/CountryDaoImpl.java [hunk 11]
```diff
@@ -12,11 +15,35 @@ public CountryDaoImpl(SessionFactory sessionFactory) {
...
     @Override
     public Country add(Country country) {
-        return null;
+        Session session = null;
+        Transaction transaction = null;
+
+        try {
+            session = factory.openSession();
+            transaction = session.beginTransaction();
+            session.save(country);
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/hibernate/relations/dao/impl/CountryDaoImpl.java:24]: See comments in ActorDaoImpl and make changes here also

File: src/main/java/mate/academy/hibernate/relations/dao/impl/MovieDaoImpl.java [hunk 12]
```diff
@@ -12,11 +15,35 @@ public MovieDaoImpl(SessionFactory sessionFactory) {
...
     @Override
     public Movie add(Movie movie) {
-        return null;
+        Session session = null;
+        Transaction transaction = null;
+
+        try {
+            session = factory.openSession();
+            transaction = session.beginTransaction();
+            session.save(movie);
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/hibernate/relations/dao/impl/MovieDaoImpl.java:24]: See comments in ActorDaoImpl and make changes here also

File: src/main/java/mate/academy/hibernate/relations/service/impl/ActorServiceImpl.java [hunk 13]
```diff
@@ -1,16 +1,24 @@
...
     public Actor add(Actor actor) {
-        return null;
+        return actorDao.add(actor);
     }
 
     @Override
     public Actor get(Long id) {
-        return null;
+        return actorDao.get(id).orElseThrow(()
+                -> new RuntimeException("No actor found with id: " + id));
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/hibernate/relations/service/impl/ActorServiceImpl.java:22]:
```suggestion
                -> new DataProcessingException("No actor found with id: " + id));
```

File: src/main/java/mate/academy/hibernate/relations/service/impl/CountryServiceImpl.java [hunk 14]
```diff
@@ -1,16 +1,24 @@
...
     public Country add(Country country) {
-        return null;
+        return countryDao.add(country);
     }
 
     @Override
     public Country get(Long id) {
-        return null;
+        return countryDao.get(id)
+                .orElseThrow(() -> new RuntimeException("Country not found with id: " + id));
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/hibernate/relations/service/impl/CountryServiceImpl.java:22]: The same

File: src/main/java/mate/academy/hibernate/relations/service/impl/MovieServiceImpl.java [hunk 15]
```diff
@@ -1,16 +1,24 @@
...
     public Movie add(Movie movie) {
-        return null;
+        return movieDao.add(movie);
     }
 
     @Override
     public Movie get(Long id) {
-        return null;
+        return movieDao.get(id).orElseThrow(()
+                -> new RuntimeException("Movie not found with id: " + id));
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/hibernate/relations/service/impl/MovieServiceImpl.java:22]: The same

**Review by Elena-Bruyako — CHANGES_REQUESTED** [review]

**Review by Sanderfuck — APPROVED** [review]

**Review by MentorMateAcademy — APPROVED** [review]

**Pull request closed without merging**